rethinkdb/handshake.py

```diff
@@ -35,306 +35,328 @@
-            self.cache[key] = val
+    @staticmethod
+    def _get_pbkdf2_hmac():
+        """
+        Get the pbkdf2_hmac function from hashlib if package contains it, else get
+        our own function. Please note that hashlib contains this function only for
+        Python 2.7.8+ and 3.4+.
+        """
+
+        return getattr(hashlib, 'pbkdf2_hmac', pbkdf2_hmac)
+
+    def _next_state(self):
+        """
+        Increase the state counter.
+        """
+
+        self._state += 1
+
+    def _decode_json_response(self, response, with_utf8=False):
+        """
+        Get decoded json response from response.
+
+        :param response: Response from the database
+        :param with_utf8: UTF-8 decode response before json decoding
+        :raises: ReqlDriverError | ReqlAuthError
+        :return: Json decoded response of the original response
+        """
+
+        if with_utf8:
+            response = decode_utf8(response)
+
+        json_response = self._json_decoder.decode(response)
+
+        if not json_response.get('success'):
+            if 10 <= json_response['error_code'] <= 20:
+                raise ReqlAuthError(json_response['error'], self._host, self._port)
+
+            raise ReqlDriverError(json_response['error'])
+
+        return json_response
+
+    def _init_connection(self, response):
+        """
+        Prepare initial connection message. We send the version as well as the initial
+        JSON as an optimization.
+
+        :param response: Response from the database
+        :raises: ReqlDriverError
+        :return: Initial message which will be sent to the DB
+        """
+
+        if response is not None:
+            raise ReqlDriverError('Unexpected response')
+
+        self._r = base64.standard_b64encode(bytes(bytearray(SystemRandom().getrandbits(8) for i in range(18))))
+        self._first_client_message = b'n={username},r={r}'.format(username=self._username, r=self._r)
+
+        initial_message = b'{pack}{message}\0'.format(
+            pack=struct.pack('<L', self.VERSION),
+            message=self._json_encoder.encode({
+                'protocol_version': self._protocol_version,
+                'authentication_method': 'SCRAM-SHA-256',
+                'authentication': b'n,,{first_message}'.format(first_message=self._first_client_message).decode('ascii')
+            }).encode('utf-8')
+        )
+
+        self._next_state()
+        return initial_message
+
+    def _read_response(self, response):
+        """
+        Read response of the server. Due to we've already sent the initial JSON, and only support a single
+        protocol version at the moment thus we simply read the next response and return an empty string as a
+        message.
+
+        :param response: Response from the database
+        :raises: ReqlDriverError | ReqlAuthError
+        :return: An empty string
+        """
+
+        if response.startswith('ERROR'):
+            raise ReqlDriverError(
```

Comment:
This might be good to raise on any failure to decode JSON.

Reply:
Is there any other scenario when the server can return `ERROR <Message>`? Or you mean raise this specific `ReqlDriverError`?

Reply:
The `ERROR` part is specific to an old DB connection. I'm just thinking of in general, if the JSON is malformed or incomplete it would be useful to raise an error with the raw output.